crates/knope-versioning/src/versioned_file/text_file.rs

```diff
@@ -0,0 +1,261 @@
+#[cfg(feature = "miette")]
+use miette::Diagnostic;
+use regex::Regex;
+use relative_path::RelativePathBuf;
+use thiserror::Error;
+
+use crate::{Action, semver::Version};
+
+#[derive(Clone, Debug)]
+pub struct TextFile {
+    pub(super) path: RelativePathBuf,
+    content: String,
+    pattern: String,
+    diff: Option<String>,
+}
+
+impl TextFile {
+    /// Creates a new TextFile with the given pattern.
+    ///
+    /// # Errors
+    ///
+    /// If the regex pattern is invalid
+    pub fn new(path: RelativePathBuf, content: String, pattern: String) -> Result<Self, Error> {
+        // Validate the regex pattern
+        Regex::new(&pattern).map_err(|source| Error::InvalidPattern {
+            pattern: pattern.clone(),
+            path: path.clone(),
+            source,
+        })?;
+
+        Ok(Self {
+            path,
+            content,
+            pattern,
+            diff: None,
+        })
+    }
+
+    /// Get the current version from the file using the regex pattern.
+    ///
+    /// # Errors
+    ///
+    /// If the pattern doesn't match or the matched version is invalid
+    pub(super) fn get_version(&self) -> Result<Version, Error> {
+        let re = Regex::new(&self.pattern).map_err(|source| Error::InvalidPattern {
+            pattern: self.pattern.clone(),
+            path: self.path.clone(),
+            source,
+        })?;
+
+        let caps = re.captures(&self.content).ok_or_else(|| Error::NoMatch {
+            pattern: self.pattern.clone(),
+            path: self.path.clone(),
+        })?;
+
+        // Try to get the first capture group, or use the full match if no groups
+        let version_str = caps
+            .get(1)
```

Comment:
I think we want all named groups named "version", not just whatever the first group happens to be.